File: adapters/sveltekit/src/server/github.ts
// Optional GitHub Issues sync.
//
// CRITICAL: GitHub sync failures MUST NOT cause the intake response to be
// non-2xx. Log the failure and return null. See PROTOCOL.md
// § "Failure modes that MUST NOT yield non-2xx".

import type { GitHubSyncOptions, BugReportDetail } from './types.js';

export interface GitHubIssueRef {
  url: string;
  number: number;
}

/**
 * Best-effort sync to GitHub Issues. Returns null on any failure.
 * Errors are written to console.warn for operator visibility.
 */
export async function syncToGitHubIssues(
  detail: BugReportDetail,
  opts: GitHubSyncOptions
): Promise<GitHubIssueRef | null> {
  if (!opts.enabled) return null;

  const apiBase = opts.apiBase ?? 'https://api.github.com';
  const url = `${apiBase}/repos/${opts.repo}/issues`;

  const body = {
    title: `[bug-fab] ${detail.title}`,
    body: buildIssueBody(detail),
    labels: opts.labels ?? ['bug-fab']
  };

  try {
    const resp = await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${opts.pat}`,
        Accept: 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(body)
    });

    if (!resp.ok) {
      console.warn(`[bug-fab] GitHub sync failed: ${resp.status} ${resp.statusText}`);
      return null;
    }

    const issue = (await resp.json()) as { html_url?: string; number?: number };
    if (!issue.html_url || typeof issue.number !== 'number') {
      console.warn('[bug-fab] GitHub sync returned malformed issue payload');
      return null;
    }
    return { url: issue.html_url, number: issue.number };
  } catch (err) {
    console.warn(`[bug-fab] GitHub sync error: ${err instanceof Error ? err.message : String(err)}`);
    return null;
  }
}

/**
 * Best-effort transition of the linked GitHub issue when status updates.
 * fixed/closed → close issue; open/investigating → reopen.
 */
export async function syncStatusToGitHub(
  detail: BugReportDetail,
  opts: GitHubSyncOptions
): Promise<void> {
  if (!opts.enabled || !detail.github_issue_number) return;
  const apiBase = opts.apiBase ?? 'https://api.github.com';
  const url = `${apiBase}/repos/${opts.repo}/issues/${detail.github_issue_number}`;
  const state = detail.status === 'fixed' || detail.status === 'closed' ? 'closed' : 'open';
  try {
    const resp = await fetch(url, {
      method: 'PATCH',
      headers: {
        Authorization: `Bearer ${opts.pat}`,
        Accept: 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ state })
    });
    if (!resp.ok) {
      console.warn(`[bug-fab] GitHub status sync failed: ${resp.status} ${resp.statusText}`);
    }
  } catch (err) {
    console.warn(`[bug-fab] GitHub status sync error: ${err instanceof Error ? err.message : String(err)}`);
  }
}

function buildIssueBody(detail: BugReportDetail): string {
  const ctx = detail.context ?? {};
  const lines: string[] = [
    `**Bug-Fab report:** \`${detail.id}\``,
    `**Severity:** ${detail.severity}`,
    `**Status:** ${detail.status}`,
    `**Reported at:** ${detail.created_at}`,
    ''
  ];
  if (detail.description) {
    lines.push('### Description', detail.description, '');
  }
  if (detail.expected_behavior) {
    lines.push('### Expected behavior', detail.expected_behavior, '');
  }
  if (ctx.url) {
    lines.push(`**Page:** ${ctx.url}`);
  }
  if (ctx.app_version) {
    lines.push(`**App version:** ${ctx.app_version}`);
  }
  if (detail.environment) {
    lines.push(`**Environment:** ${detail.environment}`);
  }
  return lines.join('\n');
}

File: adapters/rails/lib/bug_fab/github.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

module BugFab
  # Best-effort GitHub Issues sync.
  #
  # Per `repo/docs/PROTOCOL.md` § "Failure modes that MUST NOT yield non-2xx",
  # GitHub sync failures during intake or status update MUST be logged and
  # MUST NOT cause the response to be non-2xx. Every method in this module
  # rescues StandardError and returns nil on failure.
  #
  # No background workers (Sidekiq/ActiveJob): per the adapter brief, sync
  # is synchronous best-effort. Consumers wanting async sync can monkey-patch
  # this module to enqueue a job and return immediately, but Bug-Fab itself
  # adds no such dependency.
  module GitHub
    module_function

    # Create an issue from a stored bug-report detail hash. Returns
    # `[issue_number, issue_url]` on success, `[nil, nil]` on any failure.
    def create_issue(detail)
      config = BugFab.configuration
      return [nil, nil] unless config.github_configured?

      uri = URI.join(config.github_api_base + "/", "repos/#{config.github_repo}/issues")
      body = {
        title: "[Bug-Fab] #{detail[:title]}",
        body: format_body(detail),
        labels: build_labels(detail)
      }

      response = post_json(uri, body)
      return [nil, nil] unless response.is_a?(Net::HTTPSuccess)

      parsed = JSON.parse(response.body)
      [parsed["number"], parsed["html_url"]]
    rescue StandardError => e
      logger.warn("[bug_fab] github create_issue failed: #{e.class}: #{e.message}")
      [nil, nil]
    end

    # Mirror a status change onto a linked GitHub issue. fixed/closed → close;
    # open/investigating → reopen. Returns true on success, false otherwise.
    def sync_state(issue_number, status)
      config = BugFab.configuration
      return false unless config.github_configured? && issue_number

      target = case status
               when "fixed", "closed" then "closed"
               when "open", "investigating" then "open"
               else
                 return false
               end

      uri = URI.join(config.github_api_base + "/", "repos/#{config.github_repo}/issues/#{issue_number}")
      response = patch_json(uri, { state: target })
      response.is_a?(Net::HTTPSuccess)
    rescue StandardError => e
      logger.warn("[bug_fab] github sync_state failed: #{e.class}: #{e.message}")
      false
    end

    # Format the issue body. Pure presentation; not authoritative.
    #
    # Key style is load-bearing here. `BugReport#to_detail` returns a hash with
    # SYMBOL keys at the top level, whose `:context` value is the raw metadata
    # sub-hash and therefore has STRING keys. `show.html.erb` already reads it
    # that way (`detail.dig(:context, "url")`).
    #
    # This method used to read string keys throughout (`detail['severity']`,
    # `detail.dig('context', 'module')`), so every interpolation resolved to
    # nil: issues were titled "[Bug-Fab] " with an empty body and only the
    # generic `bug-fab` label. The POST succeeded, so it failed silently. Note
    # `:module` is top-level on the detail hash, not nested under `:context`.
    def format_body(detail)
      context = detail[:context] || {}
      reporter = detail[:reporter] || {}
      # `to_detail` coerces reporter fields with `.to_s`, so absent ones are
      # "" rather than nil -- `||` would happily pick the empty string.
      who = reporter[:email].presence || reporter[:name].presence || "anonymous"

      <<~MD
        **Severity:** #{detail[:severity]}
        **Module:** #{detail[:module]}
        **Reporter:** #{who}
        **App version:** #{context['app_version']}
        **Environment:** #{detail[:environment]}
        **URL:** #{context['url']}

        ---

        #{detail[:description]}

        **Expected:** #{detail[:expected_behavior]}
      MD
    end

    def build_labels(detail)
      labels = ["bug-fab"]
      labels << "severity:#{detail[:severity]}" if detail[:severity].present?
      labels << detail[:report_type] if detail[:report_type].present?
      labels.compact
    end

    def post_json(uri, body)
      request = Net::HTTP::Post.new(uri)
      apply_headers(request)
      request.body = body.to_json
      perform(uri, request)
    end

    def patch_json(uri, body)
      request = Net::HTTP::Patch.new(uri)
      apply_headers(request)
      request.body = body.to_json
      perform(uri, request)
    end

    def apply_headers(request)
      pat = BugFab.configuration.github_pat
      request["Authorization"] = "token #{pat}"
      request["Accept"]        = "application/vnd.github+json"
      request["Content-Type"]  = "application/json"
      request["User-Agent"]    = "bug_fab-rails/#{BugFab::VERSION}"
    end

    def perform(uri, request)
      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https", open_timeout: 5, read_timeout: 10) do |http|
        http.request(request)
      end
    end

    def logger
      defined?(Rails) ? Rails.logger : Logger.new($stdout)
    end
  end
end
